refactor: Improve assumeutxo state representation

ryanofsky opened the pull request:
This PR contains the first part of #28608, which tries to make assumeutxo code more maintainable, and improve it by not locking `cs_main` for a long time when the snapshot block is connected, and by deleting the snapshot validation chainstate when it is no longer used, instead of waiting until the next restart.

The changes in this PR are just refactoring. They make `Chainstate` objects self-contained, so for example, it is possible to determine what blocks to connect to a chainstate without querying `ChainstateManager`, and to determine whether a Chainstate is validated without basing it on inferences like `&cs != &ActiveChainstate()` or `GetAll().size() == 1`.

The PR also tries to make assumeutxo terminology less confusing, using "current chainstate" to refer to the chainstate targeting the current network tip, and "historical chainstate" to refer to the chainstate downloading old blocks and validating the assumeutxo snapshot. It removes uses of the terms "active chainstate," "usable chainstate," "disabled chainstate," "ibd chainstate," and "snapshot chainstate" which are confusing for various reasons.

ryanofsky committed:
refactor: Add Chainstate::m_target_blockhash member

Make Chainstate objects aware of what block they are targeting. This makes
Chainstate objects more self contained, so it's possible for validation code to
look at one Chainstate object and know what blocks to connect to it without
needing to consider global validation state or look at other Chainstate
objects.

The motivation for this change is to make validation and networking code more
readable, so understanding it just requires knowing about chains and blocks,
not reasoning about assumeutxo download states. This change also enables
simplifications to the ChainstateManager interface in subsequent commits, and
could make it easier to implement new features like creating new Chainstate
objects to generate UTXO snapshots or index UTXO data.

Note that behavior of the MaybeCompleteSnapshotValidation function is not
changing here but some checks that were previously impossible to trigger like
the BASE_BLOCKHASH_MISMATCH case have been turned into asserts.

ryanofsky committed:
refactor: Deduplicate Chainstate activation code

Move duplicate code from ChainstateManager::ActivateSnapshot and
ChainstateManager::ActivateExistingSnapshot methods to a new
ChainstateManager::AddChainstate method.

The "AddChainstate" method name doesn't mention snapshots even though it is
only used to add snapshot chainstates now, because it becomes more generalized
in a later commit in this PR ("refactor: Add ChainstateManager::m_chainstates
member")

ryanofsky committed:
refactor: Add Chainstate::StoragePath() method

Use to simplify code determining the chainstate leveldb paths. New method is
the now the only code that needs to figure out the storage path, so the path
doesn't need to be constructed multiple places and backed out of leveldb.

ryanofsky committed:
refactor: Add ChainstateManager::ValidatedChainstate() method

ValidatedChainstate() accessor replaces GetChainstateForIndexing() with no
change in behavior.

ryanofsky committed:
refactor: Add ChainstateManager::CurrentChainstate() method

CurrentChainstate() is basically the same as ActiveChainstate() except it
requires cs_main to be locked when it is called, instead of locking cs_main
internally.

The name "current" should also be less confusing than "active" because multiple
chainstates can be active, and CurrentChainstate() returns the chainstate
targeting the current network tip, regardless of what chainstates are being
downloaded or how they are used.

ryanofsky committed:
refactor: Add ChainstateManager::m_chainstates member

Use to replace m_active_chainstate, m_ibd_chainstate, and m_snapshot_chainstate
members. This has several benefits:

- Ensures ChainstateManager treats chainstates instances equally, making
  distinctions based on their attributes, not having special cases and making
  assumptions based on their identities.

- Normalizes ChainstateManager representation so states that should be
  impossible to reach and validation code has no handling for (like
  m_snapshot_chainstate being set and m_ibd_chainstate being unset, or both
  being set but m_active_chainstate pointing to the m_ibd_chainstate) can no
  longer be represented.

- Makes ChainstateManager more extensible so new chainstates can be added for
  different purposes, like indexing or generating and validating assumeutxo
  snapshots without interrupting regular node operations. With the
  m_chainstates member, new chainstates can be added and handled without needing
  to make changes all over validation code or to copy/paste/modify the existing
  code that's been already been written to handle m_ibd_chainstate and
  m_snapshot_chainstate.

- Avoids terms that are confusing and misleading:

  - The term "active chainstate" term is confusing because multiple chainstates
    will be active and in use at the same time. Before a snapshot is validated,
    wallet code will use the snapshot chainstate, while indexes will use the IBD
    chainstate, and netorking code will use both chainstates, downloading
    snapshot blocks at higher priority, but also IBD blocks simultaneously.

  - The term "snapshot chainstate" is ambiguous because it could refer either
    to the chainstate originally loaded from a snapshot, or to the chainstate
    being used to validate a snapshot that was loaded, or to a chainstate being
    used to produce a snapshot, but it is arbitrary used to refer the first
    thing. The terms "most-work chainstate" or "assumed-valid chainstate" should
    be less ambiguous ways to refer to chainstates loaded from snapshots.

  - The term "IBD chainstate" is not just ambiguous but actively confusing
    because technically IBD ends and the node is considered synced when the
    snapshot chainstate finishes syncing, so in practice the IBD chainstate
    will mostly by synced after IBD is complete. The term "fully-validated" is
    a better way of describing the characteristics and purpose of this
    chainstate.

ryanofsky committed:
refactor: Add ChainstateManager::ActivateBestChains() method

Deduplicate code looping over chainstate objects and calling
ActivateBestChain() and avoid need for code outside ChainstateManager to use
the GetAll() method.

ryanofsky committed:
refactor: Delete ChainstateManager::IsSnapshotValidated() method

IsSnapshotValidated() is only called one place outside of tests, and is use
redundantly in some tests, asserting that a snapshot is not validated when a
snapshot chainstate does not even exist. Simplify by dropping the method and
checking Chainstate m_assumeutxo field directly.

ryanofsky committed:
refactor: Delete ChainstateManager::SnapshotBlockhash() method

SnapshotBlockhash() is only called two places outside of tests, and is used
redundantly in some tests, checking the same field as other checks. Simplify by
dropping the method and using the m_from_snapshot_blockhash field directly.

ryanofsky committed:
refactor: Delete ChainstateManager::GetAll() method

Just use m_chainstates array instead.

ryanofsky committed:
doc: Improve ChainstateManager documentation, use consistent terms

ryanofsky committed:
test: Fix broken chainstatemanager_snapshot_init check

The following test code never checked anything because the if statement was
always false:

    if (cs != &chainman_restarted.ActiveChainstate()) {
        BOOST_CHECK_EQUAL(cs->m_chain.Height(), 109);
    }

Also, the height of the background chainstate it was intending to check is 110,
not 109. Fix both problems by rewriting the check.

ryanofsky committed:
refactor: Delete ChainstateManager::IsSnapshotActive() method

IsSnapshotActive() method is only called one place outside of tests and
asserts, and is confusing because it returns true even after the snapshot is
fully validated.

The documentation which said this "implies that a background validation
chainstate is also in use" is also incorrect, because after the snapshot is
validated, the background chainstate gets disabled and IsUsable() would return
false.

ryanofsky committed:
refactor: Pass chainstate parameters to MaybeCompleteSnapshotValidation

Remove hardcoded references to m_ibd_chainstate and m_snapshot_chainstate so
MaybeCompleteSnapshotValidation function can be simpler and focus on validating
the snapshot without dealing with internal ChainstateManager states.

This is a step towards being able to validate the snapshot outside of
ActivateBestChain loop so cs_main is not locked for minutes when the snapshot
block is connected.

fjahr commented:
Concept ACK

fjahr reviewed (commented):
Approach ACK

I think pretty much everything here is a worth while improvement and should set up the following improvements well. I am not so deep into the current state of the kernel work though, so I can not comment on that part, for example the newly introduced typing file. I guess @TheCharlatan could give his nod to that here.

> It removes uses of the terms "active chainstate," "usable chainstate," "disabled chainstate," "ibd chainstate," and "snapshot chainstate" which are confusing for various reasons.

I think this is a bit confusing because it sounds like you are getting rid of this completely but it's actually just avoided where you touch the code, right? Cleaning the terms up everywhere can be a follow-up, I would like to try to rewrite the design doc a bit more anyway. But I think (selfishly) would like to break it up into design and usage docs because IMO that makes it way more manageable, i.e. the last commit of #29553.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/30214#pullrequestreview-2177291599

Thanks for reviewing! Will rebase and try to address comments. I'm especially interested if you have ideas on splitting up documentation and code changes into separate PRs.

> I am not so deep into the current state of the kernel work though, so I can not comment on that part, for example the newly introduced typing file. I guess @TheCharlatan could give his nod to that here.

Sure, for background kernel/types.h is analogous to node/types.h, wallet/types.h, and common/types.h. The types files are used when libraries want to define shared struct and enum types that can be used by code NOT directly depending on the library. For example the gui code uses the `wallet::isminetype` enum type, but the gui should not depend on the wallet library, so the enum is defined in a shared header `wallet/types.h`. In this PR, wallet code uses the `kernel::ChainstateRole` struct type, but the wallet code should not depend on the kernel library, so the struct is defined in a shared header `kernel/types.h`. This pattern started with the wallet library, but has been useful for node and common libraries as well, so it is natural to extend to the kernel.

A more stricter separation between libraries would be possible with a more complicated structure, but having shared types files has been a reasonable way to separate the libraries while having a 1 library = 1 directory = 1 namespace structure.

> > It removes uses of the terms "active chainstate," "usable chainstate," "disabled chainstate," "ibd chainstate," and "snapshot chainstate" which are confusing for various reasons.
>
> I think this is a bit confusing because it sounds like you are getting rid of this completely but it's actually just avoided where you touch the code, right? Cleaning the terms up everywhere can be a follow-up, I would like to try to rewrite the design doc a bit more anyway. But I think (selfishly) would like to break it up into design and usage docs because IMO that makes it way more manageable, i.e. the last commit of #29553.

Yeah this description was written aspirationally before I implemented this PR and I did not get around to removing all the old language. I'd be interested if you have specific ideas on how to break this PR up better. I definitely think it could make sense to move documentation / comment changes to a separate PR, and I hope #29553 can be reviewed more and merged soon, before this PR. (This is a good reminder for me to re-ack #29553)

ryanofsky commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -29,6 +29,17 @@ using node::BlockManager;
 using node::KernelNotifications;
 using node::SnapshotMetadata;
 
+static std::vector<Chainstate*> GetAll(const ChainstateManager& chainman)
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r1677599944

> Why do you add this function around here instead of using `m_chainstates` as well?

Dropped this function now, it was not hard to get rid of.

ryanofsky commented on src/validation.h:
```diff
@@ -544,6 +546,18 @@ class Chainstate
+    }
+
+    //! Return true if chainstate fully validated or is assumed valid.
+    bool IsValid() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r1677573399

> nit: In other places similar checks are inlined without a helper method. That might be actually a bit more clear as well.

You're right, that is clearer. Updated different places where this was used.

ryanofsky reviewed (commented):
Rebased 493d3844cfc8628fd63184fa7a657126d9b5dc95 -> 2c1d8b426fa1832fa464c12f3cae7e2976584430 ([`pr/cstate.4`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.4) -> [`pr/cstate.5`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.4-rebase..pr/cstate.5)) due to conflicts with various prs including #30267, #30388, #30395, #29996, #30406, #30320. Also made suggested simplifications.
Rebased 2c1d8b426fa1832fa464c12f3cae7e2976584430 -> 4c995842e3689974364c8c82ee220bfaa30f7f79 ([`pr/cstate.5`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.5) -> [`pr/cstate.6`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.5-rebase..pr/cstate.6)) due to conflict with #30111
Rebased 4c995842e3689974364c8c82ee220bfaa30f7f79 -> 7fd7960389254c7a2ba50614e1cbdf8911ef6a7d ([`pr/cstate.6`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.6) -> [`pr/cstate.7`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.6-rebase..pr/cstate.7)) due to conflict with #29656

fjahr commented:
Thanks for addressing my feedback @ryanofsky ! I hope we can merge #29553 soon and then I will get back to this with another full review. But it does look very good to me already.

> I'm especially interested if you have ideas on splitting up documentation and code changes into separate PRs.

I'm unsure if it's really needed because I find clean refactoring commits like those here can be reviewed quicker than other changes, so I would be fine to keep it as is just based on volume of the change. But I did think about it and if that is feasible maybe splitting along the lines of everything that touches `ChainstateRole` and the rest (in whatever order you think works better) could be an idea. If you feel like splitting it that's fine for me but I think I would prefer to keep it as one PR as I have already passed everything so I am biased towards fewer fundamental changes ;)

Ah, and since the first is a fix you could just open that as a separate PR now. That should be an easy review and merge that doesn't conflict with anything else. The docs changes it feels like it's better to keep them along the code since you need to get your head into the topic either way. I often feel like reviewing docs and code side by side challenges your understanding and might lead to higher quality review.

TheCharlatan commented:
Concept ACK

TheCharlatan commented on src/validation.cpp:
```diff
@@ -3451,10 +3484,10 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
+    // Belt-and-suspenders check that we aren't attempting to advance the
+    // chainstate past the target block.
+    if (WITH_LOCK(::cs_main, return m_target_utxohash)) {
+        LogPrintf("m_target_utxohash is set - this chainstate should not be in operation. "
```
I've always been a bit puzzled of this check. Is it really required? If so, why don't we need to check it on every iteration of the do loop further down? What if it is actually reached, will validation on that chain just always be stuck in an unrecoverable state? Wouldn't a fatal error be more appropriate then?

TheCharlatan commented:
Reading through commit a35a2a05887202de8c49727b6ea9166bf5f1a6f2 I am asking myself if with the addition of `m_target_blockhash`, the state of `m_validity` can just be computed on the fly without having to lug around and assert its state. For tracking the change in validity to `INVALID` once the tip reaches the target block the historical chainstate, could the presence of the `m_target_utxo_hash` (similar to how the belt-and-suspenders check is done later on) be checked instead? And for the current chainstate could the `ASSUMED_VALID` state be tracked with the presence of the snapshot base block and then the transition to the `VALIDATED` state by checking if the blocks up to the snapshot base block have `BLOCK_VALID_MASK`?

ryanofsky commented:
> Reading through commit [a35a2a0](https://github.com/bitcoin/bitcoin/commit/a35a2a05887202de8c49727b6ea9166bf5f1a6f2) I am asking myself if with the addition of `m_target_blockhash`, the state of `m_validity` can just be computed on the fly without having to lug around and assert its state.

I think it could probably work but two reasons not to do it might be:

(1) to avoid performance regressions in parts of the code that are assuming it is possible to determine whether a chainstate is valid or valid instantly without without iterating over the chain
(2) to make intent explicit and make code general and future proof

To explain (2) maybe it is safe right now to assume that any chain that is not invalid and not fully validated must be assumed-valid. But it's not clear that this should always be the case. Maybe a potential libbitcoinkernel application could be a fork-monitoring client that tracks multiple chains for some other reason and doesn't want to mark any of them as assume-valid. Having an explicit assume-valid state could make purpose of the chain more explicit and could let code dealing with multiple chainstates use chains based on how they are intended to be used without trying to infer intent from other attributes.

Maybe what you are are suggesting could be an improvement, through, and it's been a while since I thought about this and reasons I listed above are speculative. Could try implementing the suggestion and see if it simplifies the code.

TheCharlatan commented:
> Maybe what you are are suggesting could be an improvement, through, and it's been a while since I thought about this and reasons I listed above are speculative. Could try implementing the suggestion and see if it simplifies the code.

This is what I had in mind: https://github.com/TheCharlatan/bitcoin/commit/9c89548c670e7b807a969e13408ff03df15143a2 (also rebased this branch to get cmake) . It is very broken in its current state though. There are a few problems with it, but I guess the fundamental problem is that the current application of a snapshot is done in relation to an existing chain and its state, and not against the block index and its state.  E.g. the check for not loading a snapshot with less work only looks at the chain state, not the block index. If the chain is behind the block index, but the block index already has more fully-validated work, the snapshot application passes, but my new function would think it is already fully validated. The tests heavily rely on the current behaviour, where the best work chain is taken from the existing chainstate.

Thinking outside of this proposed change, is this really the desired behaviour though? Shouldn't the block index be our fundamental source of truth? I can think of a few scenarios outside of the tests where the chain being behind the fully-validated block index can be encountered, for example during a chainstate reindex, but I don't think they are something where snapshots should be supported.

>  (2) to make intent explicit and make code general and future proof

I would argue that adding this validity state could also make it more complicated to implement, or compose from existing functionality, parallel validation, or similar IBD speedups that might partition the chain even more.

ryanofsky commented:
Rebased 7fd7960389254c7a2ba50614e1cbdf8911ef6a7d -> 58f62c2221d82e22e7c2d9cccedd58e4390ace5a ([`pr/cstate.7`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.7) -> [`pr/cstate.8`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.7-rebase..pr/cstate.8)) due to various conflicts.

---

re: https://github.com/bitcoin/bitcoin/pull/30214#issuecomment-2720898462

> Shouldn't the block index be our fundamental source of truth?

I'm open to this idea but not sure the block index has enough information in this case to provide information efficiently since you can't know if a particular block is fully validated without iterating over the whole chain back to genesis. Also, if mulltiple chainstates might exist for different purposes, it may be useful to store information about the purpose of each chainstate and what assumptions it was created under in the chainstate itself not in blocks that it is pointing to. I think the `Chainstate::m_validity` member in this PR does both of these things in a pretty straightforward way, but if we want to replace it with something better, hopefully that should not be too hard since it is a private member, so the public Chainstate interface would not have change.

One thing this PR is trying to accomplish is making it possible determine status of `Chainstate` objects independently, without having to look at `ChainstateManager` objects and other `Chainstate` objects. But it should be fine for `Chainstate` methods to use BlockIndex flags to return their status.

> I would argue that adding this validity state could also make it more complicated to implement, or compose from existing functionality, parallel validation, or similar IBD speedups that might partition the chain even more.

I think if multiple chainstates will exist for different purposes, the chainstate objects will need to have some field describing their purpose and controlling their behavior. I can easily see the `m_validity` field not being general enough to do that, but it's more easy for me to imagine it being replaced by another field than being eliminated entirely. But again I think the fact that it's a private field should make it easier to change or eliminate.

TheCharlatan commented:
> I think if multiple chainstates will exist for different purposes, the chainstate objects will need to have some field describing their purpose and controlling their behavior. 

Yes, I think I have come to a similar conclusion, and I think the introduced reliance here on the `m_target_blockhash` and `m_from_snapshot_blockhash` are very good changes. Having some kind of "start" and "stop" blocks for a chainstate seems like something that is easy to understand and might also be useful beyond assumeutxo.

> since you can't know if a particular block is fully validated without iterating over the whole chain back to genesis

Isn't that only true when iterating over the snapshot block or when you don't know if you might come across it? My understanding is that is also what the [docs](https://github.com/bitcoin/bitcoin/blob/master/src/chain.h#L111) say about `BLOCK_VALID_CHAIN` and `BLOCK_VALID_SCRIPTS` and is also checked in `CheckBlockIndex`.

> I think if multiple chainstates will exist for different purposes, the chainstate objects will need to have some field describing their purpose and controlling their behavior. 

I would also like to see a future with more diverse modes of validation, but I feel like we already have the `ChainstateManager` class for controlling the behaviour of a `Chainstate`. In my mind a `Chainstate` should not have to know all the details of what it is being used for, or have rich descriptive state about its own behaviour. I feel like many of the self-describing behaviours, like passing the `ChainstateRole` through a notification, should actually be done by the `ChainstateManager`. I don't think the changes in this PR go diametrically against that goal, but I am also not sure yet if the changes here really make it easier to eventually go into that direction.

ryanofsky commented:
I think I agree with all that, and think chainstate objects should just contain information for chainstatemanager to deal with them, not to have "rich" data or application specific data. I also think depending on what you want to do with chainstates, block status flags may provide enough information, but they are not equivalent to the old `Chainstate::m_disabled` field or the new `Chainstate::m_validity` field replacing it, since block flags only tell you about validity back to the last snapshot block, not back to genesis block.

I guess I'm not really sure what problems you see with chainstates knowing their role or validity, but it is true they continue to know some of that information after this PR just like they knew it before. If you want to move that information elsewhere like chainstate manager or derive it from block validity flags I don't think this PR should get in the way of that, and it probably even helps a little because it narrows the definition of `ChainstateRole` to not indicate assumeutxo status but only indicate basic information about whether the chainstate is validated and what block it is targeting. The PR should also help more if you are interested in using chainstate objects for applications other than assumeutxo, because it is dropping assumeutxo-specific assumptions many places in validation and networking code and adding a generic `ChainstateManager::m_chainstates` vector that can hold arbitrary chainstates.

I'd also point out that the `Chainstate::m_validity` field only gets written in 2 places: in the `Chainstate` constructor and in the `ChainstateManager::MaybeCompleteSnapshotValidation` method, so I don't think it adds much maintenance burden even if you think other fields might be more useful.

sedited committed:
refactor: Simplify pruning functions

Move GetPruneRange from ChainstateManager to Chainstate.

sedited reviewed (commented):
Approach ACK

I am not yet convinced of the last three commits changing the code (5e22fdcdb2665c0bb5e47c53719796096afe2dd1, 58265b955a3d8622f22958943f79409129d58b36, ca48478fbfcc83cc6f8c928e57800707fe0c3b51), maybe they could do with some more motivation in the commit message? I guess the last of the commits ca48478fbfcc83cc6f8c928e57800707fe0c3b51 is kind of the result of the preceding two and it is nice not to have to construct the vector on the fly for every `GetAll()` call anymore.

There is an opportunity for also simplifying the pruning code a bit, I made a commit here https://github.com/TheCharlatan/bitcoin/commit/fe76cc3a04c94ca4fdbb9550d478909f3b13d8ec that removes the chainman arguments from the manual pruning function. I think it would fit in well with the patch set here.

ryanofsky reviewed (commented):
Rebased d09b82156ced5d543d08226e8d7b8fda2e0ec532 -> 467528960689c2913c101ef75bc833e8f04bd0f3 ([`pr/cstate.9`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.9) -> [`pr/cstate.10`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.9-rebase..pr/cstate.10)) due to conflict 31961 and implementing various suggestions.

re: https://github.com/bitcoin/bitcoin/pull/30214#pullrequestreview-2685423528

Thanks for the review! I added comments to the `m_chainstates` commits to motivate them better, also added your simplification fe76cc3a04c94ca4fdbb9550d478909f3b13d8ec and implemented other suggestions below.

mzumsande reviewed (commented):
Concept ACK
only reviewed up to 199a2eb87c3ad9382f2fe4ae97524099224a7b04 so far.

TheCharlatan commented on src/validation.h:
```diff
@@ -598,12 +606,16 @@ class Chainstate
 
+    //! Hash of the UTXO set at the target block, computed when the chainstate
+    //! reaches the target block, and null before then.
+    std::optional<AssumeutxoHash> m_target_utxohash GUARDED_BY(::cs_main);
```
I was thinking that this might be interesting to add to the getchainstates rpc call.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2004288056

Thanks, fixed commit message. Ultimately I think it would be best to  drop the `m_target_utxohash` member and just check the `ReachedTarget()` condition everywhere instead of using `m_target_utxohash.has_value()`. But this needs to be done carefully, and it might require getting rid of some "belt and suspenders" checks in the current code which are technically unnecessary but give some confidence the code works correctly.

This commit is a more of a pure refactoring and tries to keep code the working the way it does currently, but add transparency so it clearer why it works and what assumptions it is making. It's trying to replace the opaque `m_disabled` bool which could indicate either (1) that the chainstate target is invalid or (2) that a chainstate is valid but no longer in use after its utxo hash was checked. Storing the utxo hash in the latter state seemed more natural representation than adding a `bool utxohash_checked` member, but I could switch to that if preferred.

ryanofsky reviewed (commented):
Thanks for the reviews

Rebased 467528960689c2913c101ef75bc833e8f04bd0f3 -> 00c3e4db49ebd2c3bf2f3c7f2a8c9ba5dfb3c8f5 ([`pr/cstate.10`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.10) -> [`pr/cstate.11`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.10-rebase..pr/cstate.11)) due to conflict with #31519, also adding test comments.
Update 00c3e4db49ebd2c3bf2f3c7f2a8c9ba5dfb3c8f5 -> 137d5980abd49a1d2bf783cb57a9386b1941550f ([`pr/cstate.11`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.11) -> [`pr/cstate.12`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.11..pr/cstate.12)) just fixing a commit message

TheCharlatan commented on src/kernel/types.h:
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2024 The Bitcoin Core developers
```
I really hate copyright discussions, but I think it would good to at least apply it consistently in the project. Definitely not something that should be hashed out on this PR though.

TheCharlatan commented on src/validation.cpp:
```diff
@@ -6112,20 +6131,19 @@ SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation()
     const int snapshot_base_height = *Assert(this->GetSnapshotBaseHeight());
     const CBlockIndex& index_new = *Assert(m_ibd_chainstate->m_chain.Tip());
-
-    if (index_new.nHeight < snapshot_base_height) {
```
Missed this in my previous round, but is this not changing behaviour significantly?  As far as I can tell we now just check that the target blockhash is reached at some point, while this check that is removed here only skips blocks up to the snapshot height and tries to complete validation with the rest once the snapshot base height is reached. What if the snapshot block is never reached? Does background validation then just continue on indefinitely?

I might be missing something, but the way I am reading this we also don't have any tests for exercising what previously was previously the `BASE_BLOCKHASH_MISMATCH` failure case.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2016904181

Will look more closely, but I think both the old code and new code are assuming that the chainstate tip is an ancestor of the chainstate target block, or points directly to it. If that's not true I think you are right that the behavior of this function before and after this change would be different, but a lot of other things would be broken too.

ryanofsky replied:
> re: [#30214 (comment)](https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2016904181)

> What if the snapshot block is never reached? Does background validation then just continue on indefinitely?

I think you are right and there is an improvement that can be made here. The case where the validation chainstate tip somehow points to a higher block than than target block (snapshot block) is supposed to be impossible and should never occur. That case used to return `BASE_BLOCKHASH_MISMATCH`, but was changed here to trigger an Assert instead, so the code would be simpler and so we would know if a bug occurred. However, because the Assert is place on line 6173 below the SKIPPED check on line 6126 I think the assert would never trigger, and SKIPPED would always be returned instead.

There are two callers `MaybeCompleteSnapshotValidation`, one is in `ConnectTip` which was ignoring the return value so would not be affected, but the other called in `LoadChainstate` on start would be affected and could cause the error to be ignored, instead of causing a startup failure.

I think it would probably be good to add an assert to the top of this function to make sure the chain tip is really is an ancestor of the target block, to document the assumption make make it for likely for bugs to be caught.

TheCharlatan replied:
I find it a bit hard to check this, but wouldn't the assertion then be triggerable if an alternative block at that height is received? How about adding a `ReachedTargetHeight`, and use that in the `if` block? Something like:
```diff
diff --git a/src/validation.cpp b/src/validation.cpp
index 2d19ce45e5..da9efae00f 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -6091 +6091 @@ SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation(Chai
-            !validated_chainstate.ReachedTarget()) {
+            !validated_chainstate.ReachedTargetHeight()) {
@@ -6096,0 +6097 @@ SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation(Chai
+    if (!validated_chainstate.ReachedTarget()) return SnapshotCompletionResult::BASE_BLOCKHASH_MISMATCH;
diff --git a/src/validation.h b/src/validation.h
index 1b872c5af6..2d66995440 100644
--- a/src/validation.h
+++ b/src/validation.h
@@ -643,0 +644,7 @@ public:
+    //! Return true if chainstate reached target block height.
+    bool ReachedTargetHeight() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
+    {
+        const CBlockIndex* target_block{TargetBlock()};
+        return target_block && target_block->nHeight <= m_chain.Height();
+    }
+
@@ -865,0 +873,2 @@ enum class SnapshotCompletionResult {
+    BASE_BLOCKHASH_MISMATCH,
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2017635268

Since we're talking about a case that should be impossible and indicates a bug I think the best thing to do is just to add asserts where convenient to make sure it does not happen, so I updated `ReachedTarget()` to have a new assert checking that if a target block is set the chain tip must be an ancestor of it. Your suggestion to add a `ReachedTargetHeight()` method could also work to trigger the existing assert on line 6173, but I don't think it's as appealing because any check based on height rather than hashes will be less reliable.

TheCharlatan replied:
Thanks for considering! I think that your change is good too. I was trying to write a functional test to see what happens when the snapshot block is re-organized while the historical chainstate is still catching up. It seems like we already shut down the node with a fatal error in that case, because `DisconnectTip` will fail to read the historical blocks that are not synced yet. In a way that is similar behaviour to when a re-org beyond the prune height occurs. Since that is the case I agree that reaching such a state on the historical chainstate is a programming error. Should we document this fatal error case with a functional test?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2032629060

> Should we document this fatal error case with a functional test?

It could be good to have a test or improve handling of that case, but I don't think there is much of a connection between that case and the changes in this PR.

It seems to me like implicit assumption than snapshot block will not get reorged is a different than the assumption that `!TargetBlock() || TargetBlock()->GetAncestor(m_chain.Height()) == m_chain.Tip()` will be true for any chainstate. If we would like to gracefully handle the case where the snapshot block gets reorged (which would be nice IMO), I think the best way to do that would be to call `SetTargetBlock(nullptr)` to reset the target block of the validation chainstate and turn it back into a normal chainstate.

ryanofsky reviewed (commented):
Updated 137d5980abd49a1d2bf783cb57a9386b1941550f -> abe6f0642493402d858dc3aa297ec941bfeaaf87 ([`pr/cstate.12`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.12) -> [`pr/cstate.13`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.12..pr/cstate.13)) adding a new assert to deal with concern in https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2016904181 and make sure the impossible condition that would have previously triggered a BASE_BLOCKHASH_MISMATCH error will still reliably be caught, instead of potentially keeping the background chainstate around forever in a state where it is not syncing.

TheCharlatan commented on src/test/fuzz/utxo_snapshot.cpp:
```diff
@@ -160,7 +160,7 @@ void utxo_snapshot_fuzz(FuzzBufferType buffer)
             Assert(index);
             Assert(index->nTx == 0);
-            if (index->nHeight == chainman.GetSnapshotBaseHeight()) {
+            if (index->nHeight == chainman.CurrentChainstate().SnapshotBase()->nHeight) {
```
In commit 529890d836cb75175c7943ee0a79c536412d002c

I get a compilation error here. The change seems to be two commits too early, but the line also needs to be changed in this commit, so there would be some line churn.

sedited reviewed (commented):
Last changes look good, but I think the interim compilation error in the fuzz binary should be fixed. There are some formatting and include ordering issues, can you run `clang-format-diff` over the commits?

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/30214#pullrequestreview-2749329967

Thanks for the suggestion. Will try to make sure all the intermediate commits compile again and run clang-format-diff on them in the next few days.

mzumsande commented on src/kernel/types.h:
```diff
@@ -0,0 +1,30 @@
+
+namespace kernel {
+//! Information about chainstate that notifications are sent from.
+struct ChainstateRole {
```
commit 24aaf0d3a149bf532dd5f87dc58466822f847b07:
Going from a 3-state enum to a struct of two bools means that one of the 4 possible combinations (`validated = false, historical = true`) is now something users of `ChainstateRole` must consider as a theoretical possibility, although it shouldn't be possible to actually occur.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2040309650

> commit [24aaf0d](https://github.com/bitcoin/bitcoin/commit/24aaf0d3a149bf532dd5f87dc58466822f847b07): Going from a 3-state enum to a struct of two bools means that one of the 4 possible combinations (`validated = false, historical = true`) is now something users of `ChainstateRole` must consider as a theoretical possibility, although it shouldn't be possible to actually occur.

It would be more accurate to say that state _doesn't_ occur than that it _shouldn't_ occur. The point of this change is to only give clients (wallet & indexes) chainstate information they actually care about instead of making them deal with assumeutxo design and implementation details.

Specifically indexes care (for now) whether block notifications pertain to blocks that have been fully validated from genesis since the indexes don't currently use snapshot data, while wallets do work with snapshots and only care about whether notifications pertain to the most-work chainstate. So this is information the API now directly provides instead of assumeutxo states that wallets and indexes had to make assumptions and inferences about.

A (`validated = false, historical = true`) chainstate could also be possible and even useful if you had a previous utxo snapshot that you wanted to use to generate a new snapshot targetting a later block. Theoretically you'd want the node API and wallet and index code not to require any changes if a feature like this were added, which is now the case after this commit.

fjahr replied:
I'm currently not sure why we need both `ChainValidity` and `ChainstateRole`, couldn't we have one thing that serves both purposes?

mzumsande commented on src/validation.cpp:
```diff
@@ -3961,7 +3961,7 @@ void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockInd
             pindex->m_chain_tx_count = prev_tx_sum(*pindex);
             pindex->nSequenceId = nBlockSequenceId++;
-            for (Chainstate *c : GetAll()) {
+            for (const auto& c : m_chainstates) {
```
commit 46a44dcae073758d6808444ea33eb37acd0b1248
Replacing `GetAll()` by a loop over `m_chainstates`  looks a bit dubious as a refactor to me. The latter also includes successfully validated chainstates and invalid chainstates, whereas the former doesn't. I see that this is explicitly handled in some spots such as `ActivateBestChains()`, but not everywhere, so I wonder what the consequences of this are. 

In this spot, for example, it looks that we might add blocks to `setBlockIndexCandidates` for an invalid or validate chainstate. 
This should at least be discussed in the commit message in my opinion, even if the consequences wouldn't be noticable.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2040321756

> commit [46a44dc](https://github.com/bitcoin/bitcoin/commit/46a44dcae073758d6808444ea33eb37acd0b1248) Replacing `GetAll()` by a loop over `m_chainstates` looks a bit dubious as a refactor to me. The latter also includes successfully validated chainstates and invalid chainstates, whereas the former doesn't.

The point getting rid of GetAll is to make code more explicit. If you want to skip invalid chainstates, you should explicitly skip them, not rely on a `GetAll()` method that includes and excludes chainstates in a complicated and arbitrary way. (And a misleading one, if you are a naive person who thinks "All" means all.)

> I see that this is explicitly handled in some spots such as `ActivateBestChains()`, but not everywhere, so I wonder what the consequences of this are.

This is a refactoring commit, so if there are consequences this is a bug.

> In this spot, for example, it looks that we might add blocks to `setBlockIndexCandidates` for an invalid or validate chainstate. This should at least be discussed in the commit message in my opinion, even if the consequences wouldn't be noticable.

Thanks, I think it'd be good to add something like `if (Validity() = ChainValidity::INVALID) return;` to `Chainstate::TryAddBlockIndexCandidate` with a comment to make it more obvious no behavior is changing here. This could also be a future-proofing mechanism in case in the future we wanted to handle invalid snapshots in a different way that didn't trigger an immediate shutdown.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2042356811

> Thanks, I think it'd be good to add something like `if (Validity() = ChainValidity::INVALID) return;` to `Chainstate::TryAddBlockIndexCandidate` with a comment to make it more obvious no behavior is changing here.

This is now done in the latest push

ryanofsky reviewed (commented):
Thanks for the reviews! I still planning to address the fuzz/clang-tidy comments, and will also update `TryAddBlockIndexCandidate` as mentioned below to clarify the `GetAll()` removal commit.

ryanofsky committed:
refactor: Add Chainstate m_assumeutxo and m_target_utxohash members

Get rid of m_disabled/IsUsable members. Instead of marking chains disabled for
different reasons, store chainstate assumeutxo status explicitly and use that
information to determine how chains should be treated.

ryanofsky reviewed (commented):
Rebased abe6f0642493402d858dc3aa297ec941bfeaaf87 -> feaf9fc6d15a9f06470a94423bb9e167b8657652 ([`pr/cstate.13`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.13) -> [`pr/cstate.14`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.13-rebase..pr/cstate.14)) due to conflict with #31282, also running clang-tidy and fixing the early compile error in the fuzz test

---

re: https://github.com/bitcoin/bitcoin/pull/30214#pullrequestreview-2749329967

> Last changes look good, but I think the interim compilation error in the fuzz binary should be fixed. There are some formatting and include ordering issues, can you run `clang-format-diff` over the commits?

Thanks for pointing this out! These should all be fixed now

TheCharlatan commented on src/validation.cpp:
```diff
@@ -3867,23 +3895,30 @@ void Chainstate::ResetBlockFailureFlags(CBlockIndex *pindex) {
+    // Do not continue building a chainstate that is based on an invalid
+    // snapshot. This is a belt-and-suspenders type of check because if an
+    // invalid snapshot is loaded, the node will shut down to force a manual
+    // intervention. But it is good handle this case correctly regardless.
```
Nit: `s/But it is good handle/But it is good to handle`.

sedited reviewed (approved):
ACK 12f276ba2bfc106083c1615539857cdefff9794b

fjahr commented on src/validation.h:
```diff
@@ -535,11 +542,26 @@ class Chainstate
+    kernel::ChainstateRole GetRole() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+
+    //! Return whether chain is fully validated, assumed valid, or invalid.
+    ChainValidity Validity() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
This can be marked as resolved. Not sure why this comment landed in this spot here, pretty weird, but it was going in the same direction as @mzumsande 's comment here: https://github.com/bitcoin/bitcoin/pull/30214/files#r2040321756

fjahr reviewed (commented):
Leaving some comments in case you want to address some of them with the rebase. I am mostly happy but there are a few things that I need to spend some more time on.

ryanofsky commented on src/validation.h:
```diff
@@ -1106,22 +1129,19 @@ class ChainstateManager
     const CBlockIndex* GetSnapshotBaseBlock() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
+    //! Return historical chainstate targeting a specific block, if any.
+    Chainstate* HistoricalChainstate() const EXCLUSIVE_LOCKS_REQUIRED(GetMutex())
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2093147399

> Are we then commited to the 'historical' naming now with this and I can remove all the remaining ibd/background references in a follow-up? Because that naming inconsistency has always annoyed me and now we have three names with this. I think it's very confusing for newcomers as well.

I think so and the last commit "doc: Improve ChainstateManager documentation, use consistent terms" was started to do this and I should just complete it.

ryanofsky reviewed (commented):
Thanks for the reviews! I think this PR is in a pretty good state but I need to complete the last commit to make the terminology more consistent.

Rebased 12f276ba2bfc106083c1615539857cdefff9794b -> a0672af4c75b9edcf656c5dc43aca6cb930bb165 ([`pr/cstate.15`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.15) -> [`pr/cstate.16`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.15-rebase..pr/cstate.16)) due to various conflicts, implementing various suggestions.
Updated a0672af4c75b9edcf656c5dc43aca6cb930bb165 -> f6bb0803fb6d6de5bbde84343ac14df211ed17e3 ([`pr/cstate.16`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.16) -> [`pr/cstate.17`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.16..pr/cstate.17)) fixing clang-tidy error and renaming some variables for more consistent naming

TheCharlatan commented on src/node/chainstate.cpp:
```diff
@@ -168,15 +168,16 @@ ChainstateLoadResult LoadChainstate(ChainstateManager& chainman, const CacheSize
+    Chainstate* assumed_valid_cs{chainman.DetectAssumedValidChainstate()};
 
-    if (has_snapshot && options.wipe_chainstate_db) {
+    if (assumed_valid_cs && options.wipe_chainstate_db) {
```
In commit 9c0ecea4c4df01f8ed3492e9a4f538fcef1e51df

This new `assumed_valid` terminology seems confusing to me, because assume valid and assumeutxo have different semantics and different activation heights. Snapshot seemed clear to me tbh.

ryanofsky committed:
refactor: Convert ChainstateRole enum to struct

Change ChainstateRole parameter passed to wallets and indexes. Wallets and
indexes need to know whether chainstate is historical and whether it is fully
validated. They should not be aware of the assumeutxo snapshot validation
process.

ryanofsky commented:
Updated f6bb0803fb6d6de5bbde84343ac14df211ed17e3 -> e6abe3ae66fc940ab826ac4fd6908c0724870f65 ([`pr/cstate.17`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.17) -> [`pr/cstate.18`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.17..pr/cstate.18)) to clean up a few things and avoid ASSUMED_VALID / assumed_valid everywhere since that sounds too much like assumevalid (https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2154165371).

sedited reviewed (approved):
Re-ACK 48e7e82808da43a52e31fb3ff95cdf763c077c20

sedited reviewed (approved):
Re-ACK be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2

fjahr commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -489,8 +489,7 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     // Note: cs2's tip is not set when ActivateExistingSnapshot is called.
-    Chainstate& cs2 = WITH_LOCK(::cs_main,
-        return chainman.ActivateExistingSnapshot(*assumed_base->phashBlock));
+    Chainstate& cs2 = WITH_LOCK(::cs_main, return chainman.AddChainstate(std::make_unique<Chainstate>(nullptr, chainman.m_blockman, chainman, *assumed_base->phashBlock)));
```
in 1107e04328735abe636692b1a0179c6e46613d4e

micro-nit: one of these you moved over to curly brace notation but not the others.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2568629658

> micro-nit: one of these you moved over to curly brace notation but not the others.

Thanks, switched to brace notation in the other places

fjahr commented on doc/design/assumeutxo.md:
```diff
@@ -63,7 +63,7 @@ chainstate and a sync to tip begins. A new chainstate directory is created in th
 
 When this directory is present in the datadir, the snapshot chainstate will be detected
-and loaded as active on node startup (via `DetectSnapshotChainstate()`).
+and loaded as active on node startup (via `DetectAssumeutxoChainstate()`).
```
in 21d638c92f56f5ad4d32c47240700b41e8868789

nit: `DetectUnvalidatedChainstate` would have been more consistent with the rest of the naming choices in this commit

fjahr reviewed (commented):
tACK be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2

Tested this with my own Signet snapshot and with and without pruning.

Some of the renamings could be even more descriptive for my taste but this is a big step forward, so no need to address my nits unless you really want to. If others like them they could still be applied in a simple scripted-diff follow-up.

fjahr commented:
Had another thought: since there are quite a bit of small-ish helper functions refactored or replaced with different ones, it might be interesting to check a coverage report if there are any of these new functions that aren't covered and that coverage didn't go down somewhere in general. I didn't have time to look into this myself yet though.

maflcko commented on src/validation.cpp:
```diff
@@ -1956,6 +1956,29 @@ const CBlockIndex* Chainstate::SnapshotBase() const
+    return m_cached_target_block;
+}
+
+void Chainstate::SetTargetBlock(CBlockIndex* block) EXCLUSIVE_LOCKS_REQUIRED(::cs_main)
```
0cb19b78cc297ceca591ee8be6f63b4ea7b494cc: Please don't add lock annotations to the cpp file. They won't be visible anyway. You'll have to add them to the .h file.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2570730621

> Please don't add lock annotations to the cpp file

Thanks, removed. (These were in the header file too and were probably just copied and pasted.)

maflcko commented on src/validation.cpp:
```diff
@@ -6108,23 +6127,25 @@ util::Result<void> ChainstateManager::PopulateAndValidateSnapshot(
-            !m_ibd_chainstate->m_chain.Tip()) {
-       // Nothing to do - this function only applies to the background
-       // validation chainstate.
+            !m_snapshot_chainstate->m_from_snapshot_blockhash ||
```
0cb19b78cc297ceca591ee8be6f63b4ea7b494cc: Can any of the conditions actually happen?

The snapshot chainstate not having base block hash set seems like a internal code error, than something that should be checked here? So an assert seems more appropriate, it at all? Same below for the other block hash checks.

maflcko commented on src/validation.cpp:
There are still references in this commit:

```
$ git grep m_disabled  be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2~14 
be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2~14:src/validation.cpp://  (i) setting `m_disabled` immediately and ensuring all chainstate accesses go

maflcko commented on src/validation.cpp:
```diff
@@ -3907,7 +3907,7 @@ void ChainstateManager::ReceivedBlockTransactions(const CBlock& block, CBlockInd
     auto prev_tx_sum = [](CBlockIndex& block) { return block.nTx + (block.pprev ? block.pprev->m_chain_tx_count : 0); };
     if (!Assume(pindexNew->m_chain_tx_count == 0 || pindexNew->m_chain_tx_count == prev_tx_sum(*pindexNew) ||
-                pindexNew == CurrentChainstate().SnapshotBase())) {
+                std::any_of(m_chainstates.begin(), m_chainstates.end(), [&](auto& cs) EXCLUSIVE_LOCKS_REQUIRED(cs_main) { return cs->SnapshotBase() == pindexNew; }))) {
```
nit f85ff75dad4ce8712fe65a636cf36d57b4066785: use ranges::any_of, and `const auto&`?

Also, in the commit message: `is not just ambiguous because actively confusing` -> `is not just ambiguous but actively confusing`

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2596329568

> use ranges::any_of, and `const auto&`?

Make sense, applied change.

maflcko commented on src/validation.cpp:
```diff
@@ -5512,12 +5512,10 @@ void ChainstateManager::CheckBlockIndex() const
-                const bool is_active = c == &ActiveChainstate();
-                if (!CBlockIndexWorkComparator()(pindex, c->m_chain.Tip()) && !c->setBlockIndexCandidates.contains(const_cast<CBlockIndex*>(pindex))) {
+            for (const auto& c : m_chainstates) {
+                if (!CBlockIndexWorkComparator()(pindex, c->m_chain.Tip()) && c->setBlockIndexCandidates.count(const_cast<CBlockIndex*>(pindex)) == 0) {
```
nit https://github.com/bitcoin/bitcoin/commit/f85ff75dad4ce8712fe65a636cf36d57b4066785: Any reason to revert contains to count?

maflcko commented on src/validation.h:
```diff
@@ -514,6 +517,16 @@ constexpr int64_t LargeCoinsCacheThreshold(int64_t total_space) noexcept
+    VALIDATED,
+    //! Blocks after an assumetxo snapshot have been validated but the snapshot itself has not been validated.
+    UNVALIDATED,
+    //! The assumetxo snapshot failed validation.
```
```
assumetxo -> assumeutxo [spelling mistake in comments; inconsistent with the rest of the code/documentation]

maflcko commented on src/validation.h:
```diff
@@ -855,83 +897,38 @@ enum class SnapshotCompletionResult {
- *   validated by the initial block download process.
+ * Interface for managing multiple \ref Chainstate objects, where each
+ * chainstate is associated with chainstate* subdirectory in the data directory
+ * and contains database of UTXOs existing at a different point in history. (See
```
contains database -> contains a database [missing article makes the sentence ungrammatical / harder to parse]

maflcko reviewed (commented):
needs rebase, otherwise this looks good.

left some nits/questions, but those can be ignored.


review ACK be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2 💐

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2 💐
tEVAlSI74587DOpIAcL5S2nTSMPqVt5M5BH+ebh2w27hPD9PctnGNMoQfKAnL7usnU/pNoYdlI9WAiSZpquaCA==
```

</details>

maflcko commented on src/interfaces/chain.h:
```diff
@@ -391,7 +391,9 @@ class Chain
-    //! Return true if an assumed-valid chain is in use.
+    //! Return true if an assumed-valid snapshot is in use. Note that this
+    //! returns true even after the snapshot is validated, until the next node
+    //! restart.
```
nit in cb3e7af4ed04295e9928c2b60d8ab4ba64c4efd5: I wonder if this should be fixed in a follow-up, because the comment in `AttachChain` says that the loop called when `hasAssumedValidChain` is true may be slow.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2603120305

> q in [cb3e7af](https://github.com/bitcoin/bitcoin/commit/cb3e7af4ed04295e9928c2b60d8ab4ba64c4efd5): I wonder if this should be fixed in a follow-up, because the comment in `AttachChain` says that the loop called when `hasAssumedValidChain` is true may be slow.

Yes it could be nice to change this and make loading old wallets on non-pruned nodes faster, even though it's a little bit of an edge case

ryanofsky commented on src/validation.cpp:
```diff
@@ -3461,10 +3462,10 @@ bool Chainstate::ActivateBestChain(BlockValidationState& state, std::shared_ptr<
+    // Belt-and-suspenders check that we aren't attempting to advance the
+    // chainstate past the target block.
+    if (WITH_LOCK(::cs_main, return m_target_utxohash)) {
+        LogPrintf("m_target_utxohash is set - this chainstate should not be in operation. "
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2598028084

> Could use error logging for internal bugs, and possibly add an `Assume(false)` on top?

Nice, added this.

ryanofsky commented on src/validation.cpp:
```diff
@@ -6092,41 +6083,35 @@ util::Result<void> ChainstateManager::PopulateAndValidateSnapshot(
+    // If the snapshot does not need to be validated...
+    if (unvalidated_cs.m_assumeutxo != Assumeutxo::UNVALIDATED ||
+            // Or if either chainstate is unusable...
+            !unvalidated_cs.m_from_snapshot_blockhash ||
```
https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2598179336

> q: Is there a code-path where the unvalidated chainstate has no snapshot block set? 

It think it can probably happen on startup even if it shouldn't happen normally. `MaybeValidateSnapshot` is called on startup and whenever a new block is connected, and I don't think we want callers to have to check condition of the chainstates before calling it, when it can check them itself and perform the validation if appropriate, and return SKIPPED otherwise.

ryanofsky commented on src/validation.cpp:
```diff
@@ -6151,90 +6147,82 @@ SnapshotCompletionResult ChainstateManager::MaybeCompleteSnapshotValidation()
 
     const AssumeutxoData& au_data = *maybe_au_data;
-    std::optional<CCoinsStats> maybe_ibd_stats;
+    std::optional<CCoinsStats> maybe_validated_stats;
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2568652812

> nit: How about `maybe_historical_target_stats` 🙈 It's not validated until later and the maybe refers to `std::optional` afaict.

I think the name is technically correct for the reason Marco gave, but I do think it's confusing, so renamed to `validated_cs_stats`

ryanofsky commented on src/validation.h:
```diff
@@ -1140,8 +1140,11 @@ class ChainstateManager
-    //! Returns nullptr if no snapshot has been loaded.
-    const CBlockIndex* GetSnapshotBaseBlock() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    //! Return current chainstate targeting the most-work, network tip.
+    Chainstate& CurrentChainstate() const EXCLUSIVE_LOCKS_REQUIRED(GetMutex())
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2568688269

> nit: For me "current" isn't clearer than "active" and I'm very used to active, so I am kind of -0 on this particular name change. Maybe `MostWorkChainstate` or `TipChainstate` would make explicit what is currently written in the comment.

Interesting. I feel pretty strongly that "active" is a vague and misleading name because it implies other chainstates are not active, even though both chainstates are actively being validated and having blocks attached, and both are used for applications. The current chainstate is used as the source of data for wallets and RPCs, and the historical chainstate is used as the source of data for indexes. The name "current" reflects that this chainstate is targeting the latest block as opposed to a historical one.

`MostWork` could also be ok but I don't think it's as good as `Current` because `MostWork` is not the opposite of `Historical`. `MostWork` also sounds like it refers blocks already in the chainstate, when the relevant distinction is not which blocks are in the chainstate, but which block the chainstate is targeting.

ryanofsky commented on src/kernel/types.h:
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2024 The Bitcoin Core developers
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2597734916

> nit in [1d477c7](https://github.com/bitcoin/bitcoin/commit/1d477c711599b7eeac40b18a30ee9331d951a726): Could drop the year, for entirely new files?

Yes, good catch, and done in the latest push

ryanofsky commented on src/validation.cpp:
```diff
@@ -6382,21 +6371,21 @@ util::Result<void> Chainstate::InvalidateCoinsDBOnDisk()
-    m_snapshot_chainstate.reset();
+    std::unique_ptr<Chainstate> prev_chainstate{RemoveChainstate(chainstate)};
+    Chainstate& curr_chainstate{CurrentChainstate()};
+    assert(prev_chainstate->m_mempool->size() == 0);
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2569487654

> If `RemoveChainstate` returned a `nullptr` this would crash. Maybe not a scenario that can be realistically hit currently but might still be good to handle it explicitly here or by changing `RemoveChainstate`.

Thanks, added assert here since RemoveChainstate isn't asserting. It is true that this condition should never happen because this function is called with an existing chainstate reference, but maybe some future code or a test could call it incorrectly.

ryanofsky commented on src/validation.h:
```diff
@@ -1333,6 +1333,8 @@ class ChainstateManager
     //! Get range of historical blocks to download.
     std::optional<std::pair<const CBlockIndex*, const CBlockIndex*>> GetHistoricalBlockRange() const EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
 
+    util::Result<void> ActivateBestChains() LOCKS_EXCLUDED(::cs_main);
```
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2569493043

> nit: Could have added a comment

Good catch, added documentation comment

ryanofsky reviewed (commented):
Thanks for the thorough reviews! Rebased to avoid conflicts and made (minor) updates to address all the review comments.

<!-- begin push-23 -->
Rebased be1af13426e0d4c7aa1da0eb78f15cf1fb166ba2 -> 5dd7cbf675e7ce058ccb0666dabcb39a175ae83c ([`pr/cstate.22`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.22) -> [`pr/cstate.23`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.22-rebase..pr/cstate.23))<!-- end -->

maflcko commented on src/test/util/chainstate.h:
```diff
@@ -81,7 +81,7 @@ CreateAndActivateUTXOSnapshot(
             Assert(chain.LoadGenesisBlock());
             // These cache values will be corrected shortly in `MaybeRebalanceCaches`.
-            chain.InitCoinsDB(1 << 20, true, false, "");
+            chain.InitCoinsDB(1 << 20, true, false);
```
llm nit in 6c98ddc87a895d9bbc392db8fedc0e8268ff1cdd: could use named args while touching?

```
            chain.InitCoinsDB(1 << 20, /*in_memory=*/true, /*should_wipe=*/false);

maflcko reviewed (commented):
left some more nits and questions, but they can be ignored.

re-review ACK 5dd7cbf675e7ce058ccb0666dabcb39a175ae83c 🍓

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-review ACK 5dd7cbf675e7ce058ccb0666dabcb39a175ae83c 🍓
IaZbopyDWq1aMP/wH02dao+7gmjK58ndI88eFQbAAnTEhSjYE+BlIdQ6Wam8MWiXzHApxTBD8MdQga7spNLAAg==
```

</details>

maflcko commented on src/validation.cpp:
```diff
@@ -6092,41 +6083,35 @@ util::Result<void> ChainstateManager::PopulateAndValidateSnapshot(
+            validated_cs.m_assumeutxo != Assumeutxo::VALIDATED ||
+            !validated_cs.m_chain.Tip() ||
+            // Or the validated chainstate is not targeting the snapshot block...
+            !validated_cs.m_target_blockhash ||
```
re https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2604636342

i'd say assumes are fine to detect logic bugs or storage corruption, but no strong opinion, if the goal is to somehow make this more flexible for future changes

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30214#discussion_r2605695970

> i'd say assumes are fine to detect logic bugs or storage corruption, but no strong opinion, if the goal is to somehow make this more flexible for future changes

Yeah I think it needs to be up to callers to check their own expected states or to check the return status of this function. Having this function make unnecessary assumptions about particular states it thinks it will be called in seems like it could be easily break if any outside code changes

sedited reviewed (approved):
Re-ACK 5dd7cbf675e7ce058ccb0666dabcb39a175ae83c

Rebased and addressed a bunch of nits since my last review.

ryanofsky reviewed (commented):
Thanks for the reviews!

<!-- begin push-24 -->
Rebased 5dd7cbf675e7ce058ccb0666dabcb39a175ae83c -> 82be652e40ec7e1bea4b260ee804a92a3e05f496 ([`pr/cstate.23`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.23) -> [`pr/cstate.24`](https://github.com/ryanofsky/bitcoin/commits/pr/cstate.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/cstate.23-rebase..pr/cstate.24))<!-- end --> fixing conflict #32414 and also adding suggested arg name comments

sedited reviewed (approved):
Re-ACK 82be652e40ec7e1bea4b260ee804a92a3e05f496
